examples/consumer_at_least_once_example/consumer.go

```diff
@@ -0,0 +1,141 @@
+// It has no idea what you do with the message, and it’s much more free about committing offsets.
+// As far as the consumer is concerned, as soon as a message is pulled in, it’s “processed.”
+// There aren’t any easy fixes here. Fundamentally, this is a problem of weak consistency guarantees.
+// This example ilustrates how to “roll your own” at-least-once strategy that would manually
+// commit offsets after the end of the processing pipeline.
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"gopkg.in/confluentinc/confluent-kafka-go.v1/kafka"
+)
+
+func main() {
+	if len(os.Args) < 4 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <broker> <group> <topics..>\n",
+			os.Args[0])
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	broker := os.Args[1]
+	group := os.Args[2]
+	topics := os.Args[3:]
+
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)
+
+	c, err := kafka.NewConsumer(&kafka.ConfigMap{
+		"bootstrap.servers": broker,
+		// Avoid connecting to IPv6 brokers:
+		// This is needed for the ErrAllBrokersDown show-case below
+		// when using localhost brokers on OSX, since the OSX resolver
+		// will return the IPv6 addresses first.
+		// You typically don't need to specify this configuration property.
+		"broker.address.family": "v4",
+		"group.id":              group,
+		"session.timeout.ms":    6000,
+		"auto.offset.reset":     "earliest",
+
+		// The best way to achieve at least once semantics is to
+		// disable `enable.auto.offset.store`, which marks a message as eligible
+		// for commit as soon as it's delivered to the application
+		// and use `StoreOffsets` to manually indicate this instead.
+		// Leaving `enable.auto.commit` as true to avoid blocking the poll loop.
+		"enable.auto.offset.store": false,
+		"enable.auto.commit":       true,
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create consumer: %s\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Created Consumer %v\n", c)
+
+	err = c.SubscribeTopics(topics, nil)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to subscribe topics: %s\n", err)
+		os.Exit(1)
+	}
+
+	run := true
+
+	for run == true {
+		select {
+		case sig := <-sigchan:
+			fmt.Printf("Caught signal %v: terminating\n", sig)
+			run = false
+
+		default:
+			ev := c.Poll(100)
+			if ev == nil {
+				continue
+			}
+
+			switch e := ev.(type) {
+			case *kafka.Message:
+				fmt.Printf("%% Message on %s:\n%s\n",
+					e.TopicPartition, string(e.Value))
+				if e.Headers != nil {
+					fmt.Printf("%% Headers: %v\n", e.Headers)
+				}
+
+				// We can do any long running async operation here, because the next
+				// offset will not be available for commit until we call `c.StoreOffsets(...)`
+				_ = e
+				// In this example we store the next offset that will be committed
+				// to the offset store according to `auto.commit.interval.ms` or manual
+				// offset-less Commit().
+				// In this example we choose to terminate
```

Comment:
StoreOffsets() will only return an error if you pass partitions that are unknown to the consumer, but this is never the case since the partition is supplied with the Message returned from the consumer, so I think this comment and error handling here lures people into thinking this is a case that is both common and needs handling, which is not the case.

To keep things simple I recommend removing the error handling here, and instead documenting why the Offset need to be +1 (Kafka commits are for the next message to be consumed rather than the last message processed).

Reply:
Hi @edenhill , thanks for the comment!

I believe it's not a good practice to ever ignore errors. This topic is discussed [here](https://golangbot.com/error-handling/#donotignoreerrors) and [here](https://stackoverflow.com/a/31028337/3738720). Usually when there are errors ignored throughout the code there is a need to reflect on why the error was ignored and whether it would be a bug or an interpretation of the package's behavior.

Instead of removing the error handling, what do you think about removing my comments and instead adding your explanation of why this error never happen? This will bring security to developers, especially those who are not so familiar with the package.

Reply:
For example, I have a little doubt about what happens if I call the `StoreOffset` method with an old message after a repartitioning event (if this consumer is now working with another partition).

Reply:
What happens if we manually call the `CommitMessage` method, for example:
```golang
for {
			msg, err := c.ReadMessage(1000*time.Millisecond)
			if err != nil && err.(kafka.Error).Code() != kafka.ErrTimedOut {
				log.Printf("err on instance %s -> %v", instanceID, err)
			}
			if msg != nil {
				log.Printf("Message received by [CONSUMER: %s TOPIC: %s PARTITION: %d]\n", instanceID, *msg.TopicPartition.Topic, msg.TopicPartition.Partition)
				processingFn(msg)
				_, cmtErr := consumer.CommitMessage(msg)
				if cmtErr != nil {
					break
				}
			}
		}
```
This way I can only commit messages if my `processingFn` has finished processing the message successfully

Reply:
Hi @shubhang93, if you manually call the `CommitMessage` method the consumer will block the poll loop. For systems that receive a lot of throughput this is not desirable as the consumer could be processing other messages in parallel.

> the preferred way to achieve at least once semantics is with StoreOffsets to avoid blocking the poll loop as described here: [#481 (comment)](https://github.com/confluentinc/confluent-kafka-go/issues/481#issuecomment-647820887)

Reply:
Oh, makes sense. Thanks for clearing that out!

Reply:
> I recommend documenting why the Offset need to be +1 (Kafka commits are for the next message to be consumed rather than the last message processed).

@edenhill Can we create a new method `StoreMessageOffset` as described here? https://github.com/confluentinc/confluent-kafka-go/issues/350#issuecomment-648290300

Reply:
I agree with @wedneyyuri position about the error handling. If the function can return an error it should be handled. If this change was inside the package, I would say that is still nice to handle the error, but the maintainer and contributors have enough knowledge to skip the error verification or not.

But on public a public contract? I think is a really bad decision to avoid error handling.

Reply:
> I believe it's not a good practice to ever ignore errors. 

I agree, and stand corrected.